resources/aws/lib/opentelemetry/resource/detector/aws/ec2.rb

```
@@ -0,0 +1,140 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require 'net/http'
+require 'json'
+require 'opentelemetry/common'
+
+module OpenTelemetry
+  module Resource
+    module Detector
+      module AWS
+        # EC2 contains detect class method for determining EC2 resource attributes
+        module EC2
+          extend self
+
+          # EC2 metadata service endpoints and constants
+          EC2_METADATA_HOST = '169.254.169.254'
+          TOKEN_ENDPOINT = '/latest/api/token'
+          IDENTITY_DOCUMENT_ENDPOINT = '/latest/dynamic/instance-identity/document'
+          HOSTNAME_ENDPOINT = '/latest/meta-data/hostname'
+
+          TOKEN_HEADER = 'X-aws-ec2-metadata-token'
+          TOKEN_TTL_HEADER = 'X-aws-ec2-metadata-token-ttl-seconds'
+          TOKEN_TTL_VALUE = '60'
+
+          # Timeout in seconds for HTTP requests
+          HTTP_TIMEOUT = 1
+
+          def detect
+            # Placeholder for EC2 implementation
+            resource_attributes = {}
+
+            begin
+              # Get IMDSv2 token - this will fail quickly if not on EC2
+              token = fetch_token
+              return OpenTelemetry::SDK::Resources::Resource.create({}) if token.nil?
```

For [IMDSv1](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html#instance-metadata-retrieval-examples-imdsv1), token is not required to retrieve metadata. I think it's better to allow nil token. If somehow with IMDSv2, the token is still nil, then let it fail.

[opentelemetry-js](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEc2Detector.ts) seems allow both token and identity to be nil/empty (but just make sure `identity = fetch_identity_document(token) || {}`)

---

Thanks Xuan - Updated the implementation and unit tests to support both IMDSv1 and IMDSv2 cases.